chain/chunks/src/lib.rs

```diff
@@ -1188,7 +1294,85 @@ impl ShardsManager {
             return Ok(ProcessPartialEncodedChunkResult::HaveAllPartsAndReceipts(prev_block_hash));
         }
 
-        Ok(ProcessPartialEncodedChunkResult::NeedMorePartsOrReceipts(Box::new(header)))
+        match self.chunk_forwards_cache.cache_remove(&chunk_hash) {
+            None => Ok(ProcessPartialEncodedChunkResult::NeedMorePartsOrReceipts(Box::new(header))),
+
+            Some(forwarded_parts) => {
+                // We have the header now, and there were some parts we were forwarded earlier.
+                // Let's process those parts now.
+                let forwarded_chunk = PartialEncodedChunkV2 {
+                    header: header.clone(),
+                    parts: forwarded_parts.into_iter().map(|(_, part)| part).collect(),
+                    receipts: Vec::new(),
+                };
+                // Make sure we mark this chunk as being requested so that it is handled
+                // properly in the next call. Note no requests are actually sent at this point.
+                self.request_chunk_single(header, None, protocol_version);
+                self.process_partial_encoded_chunk(
+                    forwarded_chunk,
+                    chain_store,
+                    rs,
+                    protocol_version,
+                )
+            }
+        }
+    }
+
+    /// Send the parts of the partial_encoded_chunk that are owned by `self.me` to the
+    /// other validators that are tracking the shard.
+    pub fn send_partial_encoded_chunk_to_chunk_trackers(
```

Comment:
I wonder if it makes sense to isolate this logic (along with other parts relevant only to this feature) in a `chunk_forwarding` module, which is entirely behind a feature flag. It might reduce the amount of annotated code in the main pathways, and be easier to read as a result.